### ISSUE TITLE
Using TypeScript in v2.0

### ISSUE THREAD

**Issue by RunDevelopment** [PrismJS/prism#2890]
Since we'll move to ESM for v2.0, Prism will be a lot easy to understand for static analyzers such as TypeScript. I've been wondering if we should add TypeScript to Prism in some form (either using TypeScript directly (`.ts` files) or using JSDoc).

@LeaVerou @Golmote @mAAdhaTTah

**Comment by joshgoebel** [comment]
What would the objection be to JSDoc I wonder?  I suppose contributors would need to learn the lingo if they don't already, but aren't most of your contributors to grammars anyways rather than core?  JSDoc is the way we went with Highlight.js.  I wanted the live editor linting niceness (VS Code) without moving the entire codebase to TS.  We don't have enough contributors as it is and I feel like TS could narrow that field even further (vs plain JS).  Plus I've always found it super slow compiling.  I feel that we get much of the benefit of TS using only JSDoc comments.  It can be a bit "fiddly" sometimes though.

@RunDevelopment Are you a TS developer yourself for other projects?

**Comment by LeaVerou** [comment]
I agree with @joshgoebel. Also, I believe you can have a typings file even for regular JS files.

**Comment by mAAdhaTTah** [comment]
I _like_ TypeScript but don't think it's a good fit for this project. Most of the typings I expect will be expressable via JSDoc and if they aren't, we can create `.d.ts` files for them (as @LeaVerou mentions). I wouldn't want to require people to understand TS before they can contribute. We can maintain + fix the JSDocs + `.d.ts` files ourselves without pushing that burden onto contributors.

**Comment by joshgoebel** [comment]
>  I believe you can have a typings file even for regular JS files.

You can but this is what I was referring to as "fiddly"... I think we finally have ours sorted now but for a long time no one knew exactly quite to set it up properly, I got a lot of conflicting information, a lot of "just use TypeScript" comments (well intentioned but still frustrating)... and it felt a bit like being a "second class citizen" when trying to get help from TS people.  No one was ever rude or anything like that I just got the impression no one knows how this JSDoc stuff works _entirely_ cause everyone is so busy using real full TypeScript.

Until one day I found a person on the DefinitelyTyped project who seemed to have all the answers to all the questions. :-)

**Comment by RunDevelopment** [comment]
JSDoc it is then?

Having TS as a static analyzer is a huge win either way. We don't have to go TS all the way.

> Are you a TS developer yourself for other projects?

Yes. Pretty much everything I work on is written in TS.

**Comment by joshgoebel** [comment]
> The .d.ts files are less fiddly than JSDocs in my experience.

We mix both, so maybe we're the worse case scenario. :-) Though I've been told we might still be doing it wrong and should use all JSDoc and then generate our `.d.ts` file with TS... but there are types I'd much rather express in TS than JSDoc... so ugh... :-)

**Comment by RunDevelopment** [comment]
Kind of related: Do we want to publish our own type definitions too?

Right now, there is `@types/prismjs` but it's outdated and a little wrong. Maybe we could setup something that generates type definitions for us from JSDoc?

**Comment by joshgoebel** [comment]
>Right now, there is @types/prismjs but it's outdated and a little wrong. 

We took ownership of ours just to keep them up-to-date and because if we were going to maintain them anyways, why not be canonical...

> Maybe we could setup something that generates type definitions for us from JSDoc?

The something is TypeScript.  :-)  I've been told that's how to do it if everything fits into JSDoc. https://github.com/highlightjs/highlight.js/issues/3157

**Comment by anandtiwary** [comment]
I would vote for converting this into a TS lib. It is not very different in terms of implementation. Lots of projects are on TS now and if we can support it, all those downstream projects will benefit from the type safety. I am happy to contribute if this route is approved.

**Comment by joshgoebel** [comment]
>  all those downstream projects will benefit from the type safety.

To be clear this benefit does not require converting to TS.

**Comment by RunDevelopment** [comment]
Progress update: v2 is now type checked via JSDoc + `.d.ts`. 

This includes the main library (Core, languages, plugins, etc.) and tests. Our various scripts (benchmark, danger, gulp) and the scripts of the website are not yet included. They are low-priority targets for type checking anyway.

The next part will be to figure out how to generate useable type definitions for our users.

**Comment by RunDevelopment** [comment]
Update: #3687 Long live TypeScript!
